packages/micropip/micropip/micropip.py

```diff
@@ -148,53 +138,47 @@ def __init__(self):
         )
         self.installed_packages = {}
 
-    def install(
-        self,
-        requirements: Union[str, List[str]],
-        ctx=None,
-        resolve=_nullop,
-        reject=_nullop,
-    ):
-        try:
-            if ctx is None:
-                ctx = {"extra": None}
-
-            complete_ctx = dict(markers.DEFAULT_CONTEXT)
-            complete_ctx.update(ctx)
-
-            if isinstance(requirements, str):
-                requirements = [requirements]
-
-            transaction: Dict[str, Any] = {
-                "wheels": [],
-                "pyodide_packages": set(),
-                "locked": dict(self.installed_packages),
-            }
-            for requirement in requirements:
+    async def install(self, requirements: Union[str, List[str]], ctx=None):
+        if ctx is None:
+            ctx = {"extra": None}
+
+        complete_ctx = dict(markers.DEFAULT_CONTEXT)
+        complete_ctx.update(ctx)
+
+        if isinstance(requirements, str):
+            requirements = [requirements]
+
+        transaction: Dict[str, Any] = {
+            "wheels": [],
+            "pyodide_packages": set(),
+            "locked": dict(self.installed_packages),
+        }
+        requirement_promises = []
+        for requirement in requirements:
+            requirement_promises.append(
                 self.add_requirement(requirement, complete_ctx, transaction)
-        except Exception as e:
-            reject(str(e))
+            )
 
-        resolve_count = [len(transaction["wheels"])]
+        await gather(*requirement_promises)
 
-        def do_resolve(*args):
-            resolve_count[0] -= 1
-            if resolve_count[0] == 0:
-                resolve(f'Installed {", ".join(self.installed_packages.keys())}')
+        wheel_promises = []
 
         # Install built-in packages
         pyodide_packages = transaction["pyodide_packages"]
         if len(pyodide_packages):
-            resolve_count[0] += 1
+            # Note: branch never happens in out-of-browser testing because we
+            # report that all dependencies are empty.
             self.installed_packages.update(dict((k, None) for k in pyodide_packages))
-            js_pyodide.loadPackage(list(pyodide_packages)).then(do_resolve)
+            wheel_promises.append(js_pyodide.loadPackage(list(pyodide_packages)))
 
         # Now install PyPI packages
         for name, wheel, ver in transaction["wheels"]:
-            _install_wheel(name, wheel, do_resolve, reject)
+            wheel_promises.append(_install_wheel(name, wheel))
             self.installed_packages[name] = ver
+        await gather(*wheel_promises)
```

Comment:
Could we double check that nothing can go wrong when installing wheels concurrently? 
As far as I understand, we have a guarantee that wheels names are already de-duplicated by design?

Reply:
I'm pretty sure it was installing them concurrently before using callbacks, the part that was serial was fetching the json header data from pypi (and I did confirm that those calls commute with each other).